#164691215 Validate User Signup

NicholusMuwonge opened the pull request:
#### What does this PR do?

Add relevant responses to signup fields during signup process

#### Description of Task to be completed?
- on signup, all the valid fields should be filled out otherwise user will be prompted to fill them out
- user is prompted to use the proper email format.
- user is asked to make password 8 characters long and also alphanumeric
- user cant signup with the same username or email twice
- on success a dictionary object of their username and email is returned.

#### How should this be manually tested?
- one is needed to clone this repository and pull this branch
- You can test for browser functionality using postman by running your server and testing out this route. http://127.0.0.1:8000/api/users/

#### valid input may look like:
``` {
  {
  "user":{
    "username": "jeckyll",
    "email": "hyde@dark.com",
    "password": "jakejake7"
  }
} 
````
#### response should there after look like:
``` 
{
    "user": {
        "email":  "hyde@dark.com",
        "username": "jeckyll"
    }
}
```


###### Otherwise
#### example, if you ran this
``` {
  "user":{
    "username": "jeckyll",
    "email": "hyde@dark.com",
    "password": "jakejake"
  }
} 
````
response should be :  
``` {
    "errors": {
        "password": [
            "password must contain a numeric character"
        ]
    }
}
```
#### What are the relevant pivotal tracker stories?
[https://www.pivotaltracker.com/story/show/164691215](url)

reifred commented on authors/settings.py:
```diff
@@ -83,10 +84,10 @@
+        'USER': 'postgres',
+        'PASSWORD':'',
+        # 'HOST': os.environ.get('5432'),
     }
```
@NicholusMuwonge. You shouldn't be exposing sensitive data. Consider using environment variables.

NicholusMuwonge replied:
This I am Trying to implement but still not functioning. Let me fix it

reifred commented on newenv/Include/Python.h:
```diff
@@ -0,0 +1,140 @@
+#ifndef Py_PYTHON_H
```
Please consider adding your virtual environment in the .gitignore file to avoid sending it with your codebase.

NicholusMuwonge replied:
The env file Ive added to the gitignore if you check but seems not to be ignored.

richien reviewed (changes requested):
Hi Nicholus, first of all, this is good progress. I have only noticed a few potential changes besides the changes already requested by Fred, I noticed an incomplete sentence in the first bullet point of your description of the tasks completed.
 Also, in regards to how this should be tested manually, I think that we should deal with how to test this feature in the present and not in the future after merging is done. Therefore I find the references to the staging and production environments on Heroku un-useful for testing this feature. Also, I don't see how I can use unittest to test this feature, perharps you could elaborate on what you mean. 
Lastly, I also noticed that you have 5 commits for this change, the coding convention is that we should have 1 commit per feature; perhaps you could squash those commits into one. Other than that, I have tested the feature locally and it works as you have described, well done on that.

NicholusMuwonge committed:
feat(signup validator): user gets appropriate responses

- username should be atleat four characters
- usernam should have no special characters
- all the required fields shouldnt be blank
- password should be alphanumeric
- password should be atleast 8 characters long

[Delivers #164691215]

richien reviewed (changes requested):
Thank you Nicholus for implementing the feedback. I still do not see how running unit tests can help test this implementation. Perharps you should restrict the tests to using postman as that makes more sense in regards to viewing the responses returned by certain requests.

RonKbS commented on authors/apps/authentication/validation.py:
```diff
@@ -0,0 +1,20 @@
+import re
+
+
+class Validators:
```
Nice work @NicholusMuwonge.

Please consider a different approach rather than having a class with one static method. Kindly look at serializers.py for a start.

NicholusMuwonge replied:
> Kindly include a check for whether an email exists.

hey @RonKbS this is already handled in the exceptions since the email is a unique field, so when you try to signup twice, you'll get an appropriate response, hence there's a test for it too.


output should look like:
``` 
{
    "errors": {
        "email": [
            "user with this email already exists."
        ],
        "username": [
            "user with this username already exists."
        ]
    }
}
```

NicholusMuwonge replied:
Let me also change the validation to the serializers.py as recommended

RonKbS replied:
I mean exists generally and not whether it has already been used to register with our application.

RonKbS reviewed (changes requested):
Kindly include a check for whether an email exists.

reifred commented on authors/apps/authentication/serializers.py:
```diff
@@ -26,8 +29,23 @@ class Meta:
+                    ]
+                }
+                )
         return User.objects.create_user(**validated_data)
```
Thanks, @NicholusMuwonge, for the work you could as well add an extra validator to check for a username.

RonKbS replied:
@NicholusMuwonge, @reifred's concern was not adressed.
Please do so.

NicholusMuwonge replied:
I have added the username validator @reifred

RonKbS commented on tests/test_authentication.py:
```diff
@@ -134,13 +134,21 @@ def setUp(self):
         self.client = APIClient()
+    
+    @staticmethod
+    def validator(email):
```
Is there a reason for moving this method into the tests?

NicholusMuwonge replied:
I saw the test fail and I know that the email isnt being verified so I tried verifying it before it was tested . It didnt work so I was to remove it from tests once I found solution on like a final commit

NicholusMuwonge replied:
let me take it out now

reifred reviewed (approved):
Thanks for working on the changes I requested. Though work on the conflicting files for now.

NicholusMuwonge commented:
> Thanks for working on the changes I requested. Though work on the conflicting files for now.

I have fixed the conflicts

RonKbS commented:
@NicholusMuwonge, please follow this format when writing commits 👇 
<img width="561" alt="Screenshot 2019-04-08 at 08 01 44" src="https://user-images.githubusercontent.com/40039818/55699688-a48aa800-59d4-11e9-937c-24167e6084c7.png">


Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.

NicholusMuwonge commented:
> @NicholusMuwonge, please follow this format when writing commits 
> <img alt="Screenshot 2019-04-08 at 08 01 44" width="561" src="https://user-images.githubusercontent.com/40039818/55699688-a48aa800-59d4-11e9-937c-24167e6084c7.png">
> 
> Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.

Sorry about closing the PR, on renaming the branch it automatically closed.
About the commits, I write them following the convention shared it just appears different remotely

NicholusMuwonge commented:
> > @NicholusMuwonge, please follow this format when writing commits
> > <img alt="Screenshot 2019-04-08 at 08 01 44" width="561" src="https://user-images.githubusercontent.com/40039818/55699688-a48aa800-59d4-11e9-937c-24167e6084c7.png">
> > Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.
> 
> Sorry about closing the PR, on renaming the branch it automatically closed.
> About the commits, I write them following the convention shared it just appears different remotely

let me fix them now

NicholusMuwonge commented:
> > @NicholusMuwonge, please follow this format when writing commits
> > <img alt="Screenshot 2019-04-08 at 08 01 44" width="561" src="https://user-images.githubusercontent.com/40039818/55699688-a48aa800-59d4-11e9-937c-24167e6084c7.png">
> > Also, rather than closing a PR and opening another one, approach your fellow devs or me as to how to undo an additional commit.
> 
> Sorry about closing the PR, on renaming the branch it automatically closed.
> About the commits, I write them following the convention shared it just appears different remotely

I have fixed it @RonKbS

RonKbS commented on tests/test_authentication.py:
```diff
@@ -0,0 +1,467 @@
+
+
+
+class UserTestCase(TestCase):
```
Kindly seek @richien's help on having the different test cases in their own files.
In this case they can go into files that already exist (tests for serializers can be in the same files as views if they are both concerning authentication or something similar).

NicholusMuwonge replied:
@RonKbS  I have redistributed to files that test different features kindly please take a look

RonKbS commented on authors/apps/authentication/serializers.py:
```diff
@@ -28,6 +32,30 @@ class Meta:
+                )
+
+        
+        elif len(username)< 4 or not username.isalnum():
```
Please slack Trust regarding whether users can be allowed to add a character such as an underscore or a dash in their username. This .isalnum() method restricts their use but I believe they should be used. The PO is to have a final say regarding this.

NicholusMuwonge replied:
The PO says we should allow only numbers and letters

RonKbS commented on authors/apps/authentication/tests/test_register.py:
```diff
@@ -49,3 +53,37 @@ def test_api_cannot_create_a_user_with_password_lessthan_eight_characters(self):
+            { "user": user_with_a_non_numeric_password},
+            format="json")
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn("password must contain a numeric character", response.data["errors"]["password"])
```
Please make sure that lines do not exceed 79 characters. Same for lines 71, 80 and 89.

NicholusMuwonge replied:
I have fixed this

RonKbS replied:
May I know how you've fixed it because, as seen in the picture, the line is still more than 79 characters long.
<img width="852" alt="Screenshot 2019-04-10 at 14 33 32" src="https://user-images.githubusercontent.com/40039818/55875555-ffbbc680-5b9d-11e9-92b2-e45b9e013697.png">

NicholusMuwonge replied:
sorry, I hadnt then but I have now
